### PR TITLE
Add decimal mod operator

### PR DESCRIPTION
**Pull request by pramodsatya**

### CODE REVIEW

**Review by karteekmurthys — APPROVED** [review]

**Comment by pramodsatya** [comment]
Hi @majetideepak, @aditi-pandit, could you please help review this PR?

File: velox/functions/prestosql/DecimalFunctions.cpp [hunk 1]
```diff
@@ -203,6 +203,43 @@ struct DecimalDivideFunction {
...
+    auto bType = inputTypes[1];
+    auto [aPrecision, aScale] = getDecimalPrecisionScale(*aType);
+    auto [bPrecision, bScale] = getDecimalPrecisionScale(*bType);
+    aRescale_ = computeRescaleFactor(aScale, bScale);
+    bRescale_ = computeRescaleFactor(bScale, aScale);
+  }
+
+  template <typename R, typename A, typename B>
+  void call(R& out, const A& a, const B& b) {
+    R quotient;
```
Comment (aditi-pandit) [inline velox/functions/prestosql/DecimalFunctions.cpp:226]: Nit : Maybe call this variable Q for quotient (else R seems like remainder) ?

File: velox/functions/prestosql/DecimalFunctions.cpp [hunk 2]
```diff
@@ -203,6 +203,43 @@ struct DecimalDivideFunction {
...
+    out = DecimalUtil::divide<R, A, B>(
```
Comment (aditi-pandit) [inline velox/functions/prestosql/DecimalFunctions.cpp:227]: Its possible this code has changed now but did you intend to use divideWithRoundUp ?

**Review by aditi-pandit — COMMENTED** [review]
Thanks @pramodsatya

**Review by majetideepak — COMMENTED** [review]
@pramodsatya code looks good. Can you make some tests use negative values?

**Review by majetideepak — APPROVED** [review]

**Review by mbasmanova — COMMENTED** [review]
Should we update the docs https://facebookincubator.github.io/velox/functions/presto/decimal.html ?

**Comment by pramodsatya** [comment]
> Should we update the docs https://facebookincubator.github.io/velox/functions/presto/decimal.html ?

Thanks for the feedback @mbasmanova, I have updated the docs. Could you please take another look?

File: velox/docs/functions/presto/decimal.rst [hunk 3]
```diff
@@ -120,8 +120,8 @@ the inputs.
 
 	s = max(s1, s2)
 
-To support increased scale, the result precision needs to be extended by the
-difference in s1 and s.
+To support increased scale for division, the result precision needs to be extended
+by the difference in s2 and s1.
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:124]: Why did you change "difference in s1 and s" to "difference in s2 and s1"?

File: velox/docs/functions/presto/decimal.rst [hunk 4]
```diff
@@ -130,23 +130,24 @@ difference in s1 and s.
 Like in Addition, the precision of the result may exceed 38. The choices are the
 same. Presto chooses to cap p at 38 and allow runtime errors.
 
-Let’s say `a` is of type decimal(p1, s1) with unscaled value `A` and `b` is of
-type decimal(p2, s2) with unscaled value B.
+Let’s say `a` is of type decimal(p1, s1) with unscaled value A, `b` is of
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:133]: Why are you changing this sentence?

File: velox/docs/functions/presto/decimal.rst [hunk 5]
```diff
@@ -130,23 +130,24 @@ difference in s1 and s.
...
+type decimal(p2, s2) with unscaled value B, and the result `r` is of type
+decimal(p, s) with unscaled value R.
 
 ::
 
     a = A / 10^s1
     a = B / 10^s2
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:140]: While you are here, let's fix this line: a -> b

File: velox/docs/functions/presto/decimal.rst [hunk 6]
```diff
@@ -130,23 +130,24 @@ difference in s1 and s.
...
+are:
 
 ::
 
 	s = max(s1, s2)
 	p = p1 + s2 + max(0, s2 - s1)
 
-The result 'r' has 's' digits after the decimal point and unscaled value R. We
-derive the value of R as follows:
+We derive the value of R as follows:
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:150]: Why did you remove "The result 'r' has 's' digits after the decimal point and unscaled value R" ?

File: velox/docs/functions/presto/decimal.rst [hunk 7]
```diff
@@ -159,6 +160,25 @@ then divide by B and round to the nearest whole. This method works as long as
...
+    s = max(s1, s2)
+
+To compute R, Presto rescales A and B with rescale factors :code:`max(0, s2 - s1)`
+and :code:`max(0, s1 - s2)` respectively. Then R is calculated as the modulus
+value :code:`A % B`, and the result `r` has the same sign as the dividend `a`.
+Velox implementation matches Presto.
+
+::
+
+   aRescale = (A / 10^s1) * max(1, 10^(s2 - s1))
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:178]: This part of the doc is inconsistent with the rest.

File: velox/docs/functions/presto/decimal.rst [hunk 8]
```diff
@@ -159,6 +160,25 @@ then divide by B and round to the nearest whole. This method works as long as
 rescale factor does not exceed 38. If :code:`s + s2 - s1` exceeds 38, an error
 is raised.
 
+For decimal modulus operation :code:`a % b`, Presto calculates the result
+precision and scale as follows:
+
+::
+
+    p = min(p2 - s2, p1 - s1) + max(s1, s2)
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:168]: It would be nice to explain where this formula comes from. It is not obvious.

If a and b are integers and r = a % b, then r < a && r < b, hence, the number of digits needed to represent r is no more than min of number of digits needed to represent a or b. 

When a and b are decimals, we compute mod by (1) rescaling a and b to make them integers: a * 10 ^ s, b * 10^ s; (2) computing mod of integers; (3) reporting the results using s scale. Hence, the number of digits needed to represent the result is s + min(p2 - s2, p1 - s1).

Is this how it goes?

File: velox/functions/prestosql/DecimalFunctions.cpp [hunk 9]
```diff
@@ -163,6 +163,43 @@ struct DecimalDivideFunction {
...
+struct DecimalModulusFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExec);
+
+  template <typename A, typename B>
+  void initialize(
+      const std::vector<TypePtr>& inputTypes,
+      const core::QueryConfig& /*config*/,
+      A* /*a*/,
+      B* /*b*/) {
+    auto aType = inputTypes[0];
```
Comment (mbasmanova) [inline velox/functions/prestosql/DecimalFunctions.cpp:176]: const &

File: velox/functions/prestosql/DecimalFunctions.cpp [hunk 10]
```diff
@@ -163,6 +163,43 @@ struct DecimalDivideFunction {
...
+  template <typename Q, typename A, typename B>
+  void call(Q& out, const A& a, const B& b) {
+    Q quotient;
+    out = DecimalUtil::divideWithRoundUp<Q, A, B>(
+        quotient, a, b, true, aRescale_, bRescale_);
+    DecimalUtil::valueInRange(out);
+  }
+
+ private:
+  inline static uint8_t computeRescaleFactor(
```
Comment (mbasmanova) [inline velox/functions/prestosql/DecimalFunctions.cpp:193]: Let's fold this function into the caller for readability.

File: velox/functions/prestosql/DecimalFunctions.cpp [hunk 11]
```diff
@@ -163,6 +163,43 @@ struct DecimalDivideFunction {
...
+    auto [aPrecision, aScale] = getDecimalPrecisionScale(*aType);
+    auto [bPrecision, bScale] = getDecimalPrecisionScale(*bType);
+    aRescale_ = computeRescaleFactor(aScale, bScale);
+    bRescale_ = computeRescaleFactor(bScale, aScale);
+  }
+
+  template <typename Q, typename A, typename B>
+  void call(Q& out, const A& a, const B& b) {
+    Q quotient;
+    out = DecimalUtil::divideWithRoundUp<Q, A, B>(
```
Comment (mbasmanova) [inline velox/functions/prestosql/DecimalFunctions.cpp:187]: It seems wasteful to use divideWithRoundUp which produces both remainder and quotient. Can we introduce a different helper function that only returns the remainder? It is also strange to call a function named divideWithRoundUp and pass roundUp flag 'false'?

File: velox/functions/prestosql/DecimalFunctions.cpp [hunk 12]
```diff
@@ -163,6 +163,43 @@ struct DecimalDivideFunction {
...
+        quotient, a, b, true, aRescale_, bRescale_);
+    DecimalUtil::valueInRange(out);
```
Comment (mbasmanova) [inline velox/functions/prestosql/DecimalFunctions.cpp:189]: Is this needed? Can this ever fail?

File: velox/functions/prestosql/tests/DecimalArithmeticTest.cpp [hunk 13]
```diff
@@ -403,6 +403,58 @@ TEST_F(DecimalArithmeticTest, decimalDivDifferentTypes) {
        makeFlatVector<int64_t>({100, 200, -300, 400}, DECIMAL(12, 2))});
 }
 
+TEST_F(DecimalArithmeticTest, decimalMod) {
+  // short % short -> short.
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DecimalArithmeticTest.cpp:407]: There are 6 signatures being registered, but only 4 of them are tested. Let's add tests for all signatures.

**Review by mbasmanova — COMMENTED** [review]
@pramodsatya Some comments.

File: velox/functions/prestosql/DecimalFunctions.cpp [hunk 14]
```diff
@@ -163,6 +163,61 @@ struct DecimalDivideFunction {
   uint8_t aRescale_;
 };
 
+template <typename R, typename A, typename B>
+inline static R decimalModulus(A a, B b, uint8_t aRescale, uint8_t bRescale) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/DecimalFunctions.cpp:167]: This function is used only once in a one-liner caller. Let's just fold it into the caller.

File: velox/docs/functions/presto/decimal.rst [hunk 15]
```diff
@@ -103,8 +103,8 @@ exceeds 38 and cap p at 38 when s <= 38. In this case some operations are reject
 outright while others are allowed to proceed, but may produce runtime errors. Velox
 implementation matches Presto.
 
-Division
---------
+Division and Modulus
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:106]: I feels like combining division and modulus may not be ideal. These operations are quite different. For example, division result is approximate, while modulus result is exact. Would you consider adding a separate section for modulus?

File: velox/docs/functions/presto/decimal.rst [hunk 16]
```diff
@@ -159,6 +159,39 @@ then divide by B and round to the nearest whole. This method works as long as
...
+minimum number of digits needed to represent `a` or `b`, the result precision
+needs to be increased by the smaller of the differences in the precision and
+scale of either inputs. Hence the result type precision and scale are:
+
+::
+
+	s = max(s1, s2)
+    p = min(p2 - s2, p1 - s1) + max(s1, s2)
+
+Assuming `a` and `b` have unscaled values A' and B' before they are rescaled, we
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:181]: I find this paragraph confusing.  If A is unscaled value of 'a' before rescaling, then a = A / 10^s seems wrong. Do we need this paragraph at all? Consider deleting.

**Review by mbasmanova — COMMENTED** [review]
@pramodsatya Thank you for iterating. Overall looks good % a couple small comments.

File: velox/docs/functions/presto/decimal.rst (hunk 15 above)
Comment (pramodsatya) [reply velox/docs/functions/presto/decimal.rst]: Moved to separate section, thanks.

File: velox/docs/functions/presto/decimal.rst (hunk 16 above)
Comment (pramodsatya) [reply velox/docs/functions/presto/decimal.rst]: Thanks for pointing it out, deleted this para.

File: velox/docs/functions/presto/decimal.rst [hunk 17]
```diff
@@ -136,7 +136,7 @@ type decimal(p2, s2) with unscaled value B.
 ::
 
     a = A / 10^s1
-    a = B / 10^s2
+    b = B / 10^s2
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:139]: Thank you for the fix.

File: velox/docs/functions/presto/decimal.rst [hunk 18]
```diff
@@ -174,6 +174,39 @@ digits after the decimal point, hence, max(s1, s2).
...
+    s = max(s1, s2)
+    p = min(p2 - s2, p1 - s1) + max(s1, s2)
+
+To compute R, we first rescale A and B using rescale factors :code:`max(s2 - s1, 0)`
+and :code:`max(s1 - s2, 0)` respectively. Then we compute the modulus of the
+unscaled values and ensure there are 's' digits after the decimal point.
+
+::
+
+    a = (A / 10^s1) * 10^(max(s2 - s1, 0)) = (A / 10^s)
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:205]: It is confusing to reuse A and B for original and rescaled values. Perhaps, use A and B for original values and A' and B' for rescaled values. Also, `(A / 10^s1) * 10^(max(s2 - s1, 0))` is not very readable, is it?

Consider something like this:

```
To compute R, first rescale A and B to 's':

A = a * 10^s1
B = b * 10^s2

A' = a * 10^s
B' = b * 10^s

Then, compute modulus of the rescaled values:

R = A' % B' = r * 10^s

For example, given a = 12.3 and b = 1.21, s = max(1, 2) = 2 and p = min(2, 1) + s = 3.

A = 12.3 * 10 = 123
B = 1.12 * 100 = 112

A' = 1230
B' = 112

R = 1230 % 112 = 20 = 0.20 * 100
```

File: velox/docs/functions/presto/decimal.rst [hunk 19]
```diff
@@ -174,6 +174,60 @@ digits after the decimal point, hence, max(s1, s2).
...
+
+    R = A' % B' = r * 10^s
+
+For example, say `a` = 12.3 and `b` = 1.21, `r` = :code:`a % b` is calculated
+as follows:
+
+::
+
+    s = max(1, 2) = 2
+    p = min(2, 1) + s = 3.
```
Comment (mbasmanova) [inline velox/docs/functions/presto/decimal.rst:221]: nit: remove period at the end after 3

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Comment by kgpai** [comment]
@pramodsatya Can you fix the conflict ?

**Comment by pramodsatya** [comment]
> @pramodsatya Can you fix the conflict ?

Fixed, thanks for pointing it out @kgpai.

**Review by Yuhta — APPROVED** [review]
